src/prevector.h

```diff
@@ -410,17 +413,9 @@ class prevector {
         // necessary to switch to the (more efficient) directly allocated
         // representation (with capacity N and size <= N).
         iterator p = first;
-        char* endp = (char*)&(*end());
-        if (!std::is_trivially_destructible<T>::value) {
-            while (p != last) {
-                (*p).~T();
-                _size--;
-                ++p;
-            }
-        } else {
-            _size -= last - p;
-        }
-        memmove(&(*first), &(*last), endp - ((char*)(&(*last))));
+        char* endp = reinterpret_cast<char*>(end().ptr);
+        _size -= last - p;
+        memmove(first.ptr, last.ptr, endp - reinterpret_cast<char*>(last.ptr));
```

Comment:
```suggestion
        std::memmove(first.ptr, last.ptr, endp - reinterpret_cast<char*>(last.ptr));
```

Maybe switch to `std::memmove`, which comes with the documentation:

https://en.cppreference.com/w/cpp/string/byte/memmove

> If the objects are [potentially-overlapping](https://en.cppreference.com/w/cpp/language/object#Subobjects) or not [TriviallyCopyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable), the behavior of memmove is not specified and [may be undefined](http://stackoverflow.com/questions/29777492).

Reply:
Oh, if the changes are more expensive, then maybe do it in another commit or pull request.

Reply:
Does UB section of the documentation in https://en.cppreference.com/w/cpp/string/byte/memmove apply to https://en.cppreference.com/w/c/string/byte/memmove?

Reply:
I am not familiar with C, but maybe every class is trivially copyable in C?

Reply:
Right (interesting [historic context](https://www.stroustrup.com/bs_faq.html#invention) here). So IIUC the premise in the pull description only applies to the C++ memmove and it may make sense to change to it at the same time, or not at all?

Reply:
> the premise in the pull description only applies to the C++ memmove and it may make sense to change to it at the same time (or not do this change at all).

I still think it is necessary to do the change, even when we stay with `memmove`. The current implementation just does not work with complex objects. E.g. here is a small example:

```cpp
BOOST_AUTO_TEST_CASE(Erase) {
    prevector<8, std::string> pv;
    pv.push_back("a");
    pv.push_back("b");
    pv.erase(pv.begin());
    pv.push_back("X");
    std::cout << pv.front() << std::endl;
}
```

One would expect that this prints "b", as "a" was deleted and "b" should now be the first element. But l get this output:

```
Running 1 test case...
X
free(): invalid pointer
unknown location(0): fatal error: in "prevector_tests/Erase": signal: SIGABRT (application abort requested)
```

So suddenly the first element is not "b" but "X". This is due to to small string optimization, the strings point to member data, and this is simply copied. Complex objects have to be moved for this to work.

here is another motivating example with `std::swap`:

```cpp
BOOST_AUTO_TEST_CASE(NonTrivial) {
    prevector<8, std::string> a;
    prevector<8, std::string> b;
    a.push_back("a");
    b.push_back("b");

    std::cout << "a=" << a.front() << ", b=" << b.front() << std::endl;
    std::swap(a, b);
    std::cout << "a=" << a.front() << ", b=" << b.front() << std::endl;
}
```
Running that test produces this output:

```
Running 1 test case...
a=a, b=b
a=a, b=b
free(): invalid pointer
unknown location(0): fatal error: in "prevector_tests/NonTrivial": signal: SIGABRT (application abort requested)
```

So swap doesn't seem to do anything for these objects, and destructing will cause a SIGABRT. 

I think we should either restrict prevector to only work with supported types, or to correctly implement it for nontrivial types. Fixing this is quite a bit more work though, and would need a lot more testing.

Reply:
Somewhat followed up in #25172.